File: problem-3.js
var num = 600851475143;

//returns an array of the factors of the given number
var findFactors = function(x){
	var factors = [];
	for(var i = 2; i < x; i++){
		if(x % i == 0){
			factors.push(i);
			if(factors.indexOf(x/i) == -1){
				factors.push(x/i);
			}
			else{
				break;
			};
		};
	};
	return factors;
};

//a number is prime if it isn't devisible by the prime number
//that are less than its sqare root.

//returns true if the given value is prime
var isPrime = function(x){
	var prime = true;
	for(var i = 2; i<=Math.sqrt(x); i++){
		if(x % i == 0){
			prime = false;
		};
	};
	return prime;
};

//returns an array of prime numbers.
//takes an array as an argument.
var findPrime = function(a){
	var primes = [];
	for(var i = 0; i<a.length; i++){
		if(isPrime(a[i])){
			primes.push(a[i]);
		}
	};
	return primes;
};

//returns the largest of values in an array;
var findBiggest = function(a){
	var big = 0;
	for(var i = 0; i < a.length; i++){
		if(a[i] > big){
			big = a[i];
		};
	};
	return big;
};

print(findBiggest(findPrime(findFactors(num))));

File: problem-3.rb
number = 600851475143

#returns an array of the factors of the given number
def find_factors number
	factors = []
	(2..number).each do |n|
		if number % n == 0
			if factors.include? n
				break
			else
				factors<<n
				factors<<number/n
			end
		end
	end
	factors
end

#returns ana array of prime elements from a given array
def find_prime array
	primes = []
	array.each { |n| primes << n if is_prime? n }
	primes
end

#returns the largest element of an array
def find_biggest array
	biggest = 0
	array.each { |n| biggest = n if n > biggest}
	biggest
end

#returns true if a number is prime
def is_prime? number
	(2..Math.sqrt(number)).each {|n| return false if number % n == 0}
	true
end

puts find_biggest(find_prime(find_factors number))
puts "yaaay!"

